synapse/crypto/keyring.py

```diff
@@ -135,7 +135,18 @@ class KeyLookupError(ValueError):
 
 @attr.s(slots=True)
 class _FetchKeyRequest:
-    """A request for keys for a given server."""
+    """A request for keys for a given server.
+
+    We will continue to try and fetch until we have all the keys listed under
+    `key_ids` (with an appropriate `valid_until_ts` property) or we run out of
+    places to fetch keys from.
+
+    Attributes:
+        server_name: The name of the server that owns the keys.
+        minimum_valid_until_ts: The earliest timestamp at which we need the
+            keys to be valid at.
```

Comment:
this is a particularly mind-bending way of defining this, though I agree it's not easy.

```suggestion
        minimum_valid_until_ts: The timestamp which the keys must be valid until.
```